rfc/rfc-36/rfc-36.md

```diff
@@ -0,0 +1,605 @@
+
+- **timeline service**
+    -  is for timeline related requests. The basic functions are the same as the `HoodieActiveTimeline`.
+
+    -  To client, it exposes API that
+
+    - the instant CRUD
+    - get the metadata of an instant
+
+- **snapshot service**
+    -  is for file level related request. The old way is that a write client gets all files under the path by file listing, and then builds the file system view by organizing the files as file slices in file groups. After that, get the latest file slice in each file group to read and write. Snapshot service just gives the client the result of the last step.
+
+    -  To client, it exposes API that
+
+    - get the latest snapshot of a partition without multiple file version
+    - get the incremental files after a specified timestamp, for incremental reading
+    - get all files of a partition,for file cleaning
+    - update the snapshot when a new commit completes, it cannot be invoked by client directly.
+
+The specific interface is defined in the `Alpha Version` part.
+
+### Storage
+
+According to the type of the metadata stored, the storage is divided into four parts: table, partition, timeline, snapshot. Table consists of location, schema, table configurations and so on. Partition is just partition metadata. Timeline contains instant and the metadata of each instant. Snapshot has file statuses.
+
+#### Lectotype
+
+**Requirements:** support single row ACID transactions. Almost all write operations require it, like instant creation, state changing and so on.
+
+There are the candidates,
+
+**Hudi table**
+
+pros:
+
+- No external components are introduced and maintained.
+
+crons:
+
+- It is unreasonable that the storage of the hudi metastore server is a hudi table. How does the metadata of the server storage store? There is a paradox.
+
+**RDBMS**
+
+pros:
+
+- database that is suitable for structured data like metadata to store.
+
+- can describe the relation between many kinds of metadata.
+
+crons:
+
+- introduce another system to maintain.
+
+**File system**
+
+pros:
+
+- No external components are introduced and maintained.
+
+crons:
+
+- not suitable for the situation that requires high performance.
+
+- have to do extra work to support the metadata organization.
+
+**Key-value storage**
+
+pros:
+
+- database that is suitable for structured data like metadata to store.
+
+- in-memory data store so that read and write faster.
+
+crons:
+
+- introduce another system to maintain.
+
+- stroage capacity is a limitation.
+
+There are specific requirements of the metastore server in the different scenarios. Through the storage of server is pluggable, considering the general situation of disk storage, good performance of read and write, convenience of development, RDBMS may be a better one to be chosen.
+
+#### Storage Schema
```

Comment:
We need to think about the Storage schema version at the highest level and every evolution of the storage schema needs to ensure strict forwards and backwards compatibility. Think about scenarios where hudi is upgraded and then downgraded to an older version in production.

Reply:
It's a good question about the compatibility. In my opinion, scalability will brings the big changes of the storage schema. So, at first we thought db sharding could make the metastore storage scalable. Tbl_id is a suitable to be shard key. Cause shard limits the join action and only relations belong to the same tbl_id will connect together.
During the metastore evolution at ByteDance, the other reason results in schema changes can all be solved by add a new column with default value to the table. There is no compatibility problem.